### PR TITLE
Add Checkout feature

### PR DESCRIPTION
**Pull request by SAR2652**
Implemented checkout feature and corresponding tests. Added support for the ```active``` field in the Inventory Model for most routes and tests. Also added condition in update and checkout URLs

### CODE REVIEW

File: service/models.py [hunk 1]
```diff
@@ -73,10 +72,13 @@ def update(self, new_data):
         Updates a Inventory to the database
         """
         self.name = new_data.name or self.name
-        self.quantity = new_data.quantity or self.quantity
+        if hasattr(new_data, 'quantity'):
+            self.quantity = new_data.quantity
```
Comment (aayushagrawal135) [inline service/models.py:76]: Is this change necessary?

File: service/models.py [hunk 2]
```diff
@@ -73,10 +72,13 @@ def update(self, new_data):
...
         self.reorder_quantity = new_data.reorder_quantity or self.reorder_quantity
         self.restock_level = new_data.restock_level or self.restock_level
         self.updated_at = datetime.utcnow()
+        if new_data.active is False:
+            self.active = new_data.active
```
Comment (aayushagrawal135) [inline service/models.py:81]: Why do we reassign only on cases of `False`?

File: service/models.py [hunk 3]
```diff
@@ -144,6 +147,12 @@ def deserialize_util(self, data):
             else:
                 raise TypeError
 
+        if data.get("active"):
+            if isinstance(data.get("active"), bool):
+                setattr(self, "active", data.get("active"))
+            else:
+                raise TypeError
+
```
Comment (aayushagrawal135) [inline service/models.py:155]: Try to rebase the branch over #68's branch, may get merge conflicts later.

File: service/routes.py [hunk 4]
```diff
@@ -140,6 +140,40 @@ def update_inventory_records(product_id):
...
+        new_record.quantity = 0
+        new_record.active = False
+        new_record.checkout_flag = True
+    else:
+        new_record.quantity = existing_record.quantity - ordered_quantity
+        app.logger.info("Executed till here")
+
+    existing_record.update(new_record)
+    return jsonify(existing_record.serialize()), status.HTTP_200_OK
+
```
Comment (aayushagrawal135) [inline service/routes.py:176]: Move the logic of checkout to the models file to make routes lighter. Can call that logic from the routes.

File: tests/factories.py [hunk 5]
```diff
@@ -20,3 +20,4 @@ class Meta:  # pylint: disable=too-few-public-methods
     quantity = FuzzyChoice(choices=[10, 15, 20])
     reorder_quantity = FuzzyChoice(choices=[10, 15, 20])
     restock_level = FuzzyChoice(choices=[1, 2, 3])
+    active = True
```
Comment (aayushagrawal135) [inline tests/factories.py:23]: Rebase over #68.

File: service/models.py (hunk 1 above)
Comment (SAR2652) [reply service/models.py]: Updated in next commit

File: service/models.py (hunk 2 above)
Comment (SAR2652) [reply service/models.py:82]: Because the OR statement written previously was unable to assign false if ```new_data.active``` is false

File: service/models.py (hunk 3 above)
Comment (SAR2652) [reply service/models.py:156]: I am expecting merge conflicts. Lets see

File: tests/factories.py (hunk 5 above)
Comment (SAR2652) [reply tests/factories.py:23]: Yes

**Comment by sakshambassi** [comment]
Please edit the PR title to something like Add Checkout feature for ...
This will help easily track PRs in the future or otherwise.

File: service/models.py [hunk 6]
```diff
@@ -73,10 +72,14 @@ def update(self, new_data):
         Updates a Inventory to the database
         """
         self.name = new_data.name or self.name
-        self.quantity = new_data.quantity or self.quantity
+        # if hasattr(new_data, 'quantity'):
+        #     self.quantity = new_data.quantity
+        self.quantity = new_data.quantity
```
Comment (aayushagrawal135) [inline service/models.py:77]: Remove commented code.

File: tests/test_routes.py [hunk 7]
```diff
@@ -334,9 +333,99 @@ def test_method_not_allowed(self):
...
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+        
+
+    
+        
+
+
+
+
+
```
Comment (aayushagrawal135) [inline tests/test_routes.py:431]: Leave only 1 empty line at the end.

File: service/routes.py [hunk 8]
```diff
@@ -140,6 +140,38 @@ def update_inventory_records(product_id):
     existing_record.update(new_record)
     return jsonify(existing_record.serialize()), status.HTTP_200_OK
 
+
+@app.route("/inventory/checkout/<int:product_id>", methods=["PUT"])
```
Comment (richa15997) [inline service/routes.py:144]: add condition in URL

File: tests/test_routes.py [hunk 9]
```diff
@@ -334,9 +333,99 @@ def test_method_not_allowed(self):
...
+        logging.debug('Created record, %s', Inventory().deserialize(response.get_json()))
+
+        # Get record as JSON
+        data = response.get_json()
+        request_dict = dict()
+        request_dict['product_id'] = data['product_id']
+        request_dict['condition'] = data['condition']
+        request_dict['ordered_quantity'] = data['quantity'] + 1
+        logging.debug(type(data['condition']))
+        response = self.client.put(f"{BASE_URL}/checkout/{request_dict['product_id']}",
```
Comment (richa15997) [inline tests/test_routes.py:402]: pass condition in URL for all tests

File: tests/test_routes.py [hunk 10]
```diff
@@ -334,9 +333,99 @@ def test_method_not_allowed(self):
...
+        data = response.get_json()
+        request_dict = dict()
+        request_dict['product_id'] = data['product_id'] + 1
+        request_dict['condition'] = data['condition']
+        request_dict['ordered_quantity'] = data['quantity'] + 1
+        logging.debug(type(data['condition']))
+        response = self.client.put(f"{BASE_URL}/checkout/{request_dict['product_id']}",
+                                   json=request_dict)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
```
Comment (richa15997) [inline tests/test_routes.py:423]: run pylons

File: service/routes.py [hunk 11]
```diff
@@ -140,6 +140,38 @@ def update_inventory_records(product_id):
...
+
+    condition = Inventory.Condition(data['condition']).name
+    app.logger.info(type(condition))
+    app.logger.info(condition)
+    ordered_quantity = data['ordered_quantity']
+    existing_record = Inventory.find((product_id, condition))
+    app.logger.info(existing_record)
+
+    if not existing_record:
+        abort(status.HTTP_404_NOT_FOUND, f"Product with id '{product_id}' was not found.")
```
Comment (richa15997) [inline service/routes.py:158]: add given condition in the string

**Review by richa15997 — APPROVED** [review]

File: service/models.py [hunk 12]
```diff
@@ -73,10 +72,14 @@ def update(self, new_data):
         Updates a Inventory to the database
         """
         self.name = new_data.name or self.name
-        self.quantity = new_data.quantity or self.quantity
+        # if hasattr(new_data, 'quantity'):
```
Comment (sakshambassi) [inline service/models.py:75]: if this is not required, you can delete it in later commits

File: service/routes.py [hunk 13]
```diff
@@ -142,7 +142,7 @@ def list_inventory_records():
     return jsonify(results), status.HTTP_200_OK
 
 
-@app.route("/inventory/<product_id>", methods=["DELETE"])
+@app.route("/inventory/<int:product_id>", methods=["DELETE"])
```
Comment (sakshambassi) [inline service/routes.py]: why was this change done?

File: service/routes.py [hunk 14]
```diff
@@ -174,6 +176,40 @@ def update_inventory_records(product_id):
...
     return jsonify(existing_record.serialize()), status.HTTP_200_OK
 
+
+@app.route("/inventory/checkout/<int:product_id>/<condition>/", methods=["PUT"])
+@app.route("/inventory/checkout/<int:product_id>/<condition>", methods=["PUT"])
+def checkout_quantity(product_id, condition):
+    """Reduces quantity from inventory of a particular item based on the amount specified by user"""
+    app.logger.info("Reduce quantity of item based on user requirement")
+    data = request.get_json()
+    app.logger.info(type(condition))
```
Comment (sakshambassi) [inline service/routes.py:186]: either add text to logging or delete the line, this might not help in debugging.

File: service/routes.py [hunk 15]
```diff
@@ -174,6 +176,40 @@ def update_inventory_records(product_id):
...
+    # condition = Inventory.Condition(data['condition']).name
+    condition_enum = Inventory.Condition(condition).name
+    app.logger.info(type(condition))
```
Comment (sakshambassi) [inline service/routes.py:189]: same here, can remove this logging line

File: service/routes.py [hunk 16]
```diff
@@ -174,6 +176,40 @@ def update_inventory_records(product_id):
...
+    app.logger.info(condition)
```
Comment (sakshambassi) [inline service/routes.py:190]: ditto like above

File: service/routes.py [hunk 17]
```diff
@@ -174,6 +176,40 @@ def update_inventory_records(product_id):
...
+    existing_record = Inventory.find((product_id, condition_enum))
+    app.logger.info(existing_record)
+
+    if not existing_record:
+        abort(status.HTTP_404_NOT_FOUND, f"Product with id '{product_id}' was not found.")
+
+    new_record = Inventory()
+    if ordered_quantity > existing_record.quantity:
+        del new_record
+        abort(status.HTTP_405_METHOD_NOT_ALLOWED, f"Quantity specified is more than quantity of"
```
Comment (sakshambassi) [inline service/routes.py:201]: nit: f string can be made into one

File: service/models.py (hunk 12 above)
Comment (SAR2652) [reply service/models.py:75]: fixed it

File: service/routes.py (hunk 14 above)
Comment (sakshambassi) [reply service/routes.py:186]: regarding line 186

File: service/routes.py [hunk 18]
```diff
@@ -174,6 +176,35 @@ def update_inventory_records(product_id):
...
+
+@app.route("/inventory/checkout/<int:product_id>/<condition>/", methods=["PUT"])
+@app.route("/inventory/checkout/<int:product_id>/<condition>", methods=["PUT"])
+def checkout_quantity(product_id, condition):
+    """Reduces quantity from inventory of a particular item based on the amount specified by user"""
+    data = request.get_json()
+    condition_enum = Inventory.Condition(condition).name
+    ordered_quantity = data['ordered_quantity']
+    existing_record = Inventory.find((product_id, condition_enum))
+    app.logger.info(existing_record)
```
Comment (sakshambassi) [inline service/routes.py:188]: can remove this logger line

File: tests/test_models.py [hunk 19]
```diff
@@ -63,16 +63,18 @@ def test_instantiate_inventory_record(self):
         self.assertEqual(record.restock_level, 2)
 
     def test_inventory_serialize(self):
-        record = Inventory(product_id=1, name="monitor", condition=Inventory.Condition.NEW, quantity=10, reorder_quantity=20, restock_level=2, active=False)
+        record = Inventory(product_id=1, name="monitor", condition=Inventory.Condition.NEW,
+                            quantity=10, reorder_quantity=20, restock_level=2, active = True)
         actual_output = record.serialize()
+        logging.debug(actual_output)
```
Comment (sakshambassi) [inline tests/test_models.py:69]: remove logging line

File: tests/test_routes.py [hunk 20]
```diff
@@ -520,3 +522,91 @@ def test_health(self):
...
+        data = response.get_json()
+        request_dict = dict()
+        request_dict['ordered_quantity'] = data['quantity'] + 1
+        logging.debug(type(data['condition']))
+        response = self.client.put(f"{BASE_URL}/checkout/{data['product_id'] + 1}"
+                                   f"/{data['condition']}",
+                                   json=request_dict)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+        
+
```
Comment (sakshambassi) [inline tests/test_routes.py:605]: delete these extra lines

**Review by richa15997 — APPROVED** [review]

**Review by sakshambassi — APPROVED** [review]
